SCI32: palette management (full kPalCycle, partial kSetPalStyleRange and kPalFade)

csnover opened the pull request:
This PR includes implementation of the kPalCycle opcode, partial transcription of kSetPalStyleRange (commented out, since more work needs to be done in order to architect palette code to match the SCI32 architecture before it can be implemented), and some stubs for other parts of palette code from SCI32. This work is based on SQ6DOS disassembly.

There is also a small patch to scummsys.h to fix some warnings clang was emitting about redefined constants.

I’ll accept any feedback to improve the PR/fix problems in the code. Apologies that some things are interleaved into commits a little weirdly; if needed I can try to rewrite the commits to group functionality a little bit better. This was just an artefact of getting stuck and switching between multiple different features in the code.

Thanks!

bluegr commented on engines/sci/graphics/frameout.cpp:
```diff
@@ -658,6 +749,11 @@ void GfxFrameout::kernelFrameout() {
 	_palette->palVaryUpdate();
+	_palette->applyCycles();
+	_palette->applyFade();
+	// TODO: This should probably should not require screen pic invalidation
```
Wording: "This should probably not require pic invalidation"

csnover replied:
Fixed

lordhoto commented:
9d4e442ede1a29468f0fb7e94cb401e831b9605a should really not be in this PR

bluegr commented on engines/sci/engine/kgraphics32.cpp:
```diff
@@ -707,66 +743,64 @@ reg_t kPalVaryUnknown2(EngineState *s, int argc, reg_t *argv) {
+		uint16 fromColor = argv[1].toUint16();
+		uint16 toColor = argv[2].toUint16();
+		int16 direction = argv[3].toSint16();
+		uint16 delay = (argc == 4 ? 0 : argv[4].toUint16());
```
Since these 4 are the most variables that are actually used in this switch, it would be better to define them outside the switch, to avoid enclosing all of the cases in curly brackets

wjp replied:
I disagree there. Limiting scope is a good thing.

csnover replied:
For now I will keep it as-is but if you guys want to cage fight it out, let me know who wins. :)

bluegr commented on engines/sci/graphics/palette.cpp:
```diff
@@ -819,6 +824,8 @@ bool GfxPalette::palVaryLoadTargetPalette(GuiResourceId resourceId) {
 }
 
+// TODO: This code should not be using an independent timer; in SCI engine palette, varies happen in
+// Palette::UpdateForFrame which is called by GraphicsMgr::FrameOut.
```
This comment is only valid for SCI32 games

csnover replied:
Fixed

bluegr commented on engines/sci/graphics/palette.cpp:
```diff
@@ -1139,6 +1146,257 @@ void GfxPalette::unloadClut() {
+
+	_setCycleMap(fromColor, numColorsToCycle);
+
+	// TODO: Validate return value
```
Validate it... how?

csnover replied:
Clarified

bluegr commented on engines/sci/graphics/palette.h:
```diff
@@ -111,9 +159,44 @@ class GfxPalette : public Common::Serializable {
+	void cycleAllOff();
+
+	void setFade(uint8 percent, uint16 fromColor, uint16 toColor);
+	void fadeOff();
```
I've discussed these with m_kiewitz as well. Please, create a new GfxPalette32 class for SCI32 games, which inherits from GfxPalette and defines the extra methods needed for SCI32. This is a cleaner approach, as it separates the SCI32 code from the SCI16 code, and we've used it in other parts of the engine as well.

csnover replied:
> Please, create a new GfxPalette32 class for SCI32 games, which inherits from GfxPalette and defines the extra methods needed for SCI32. This is a cleaner approach, as it separates the SCI32 code from the SCI16 code, and we've used it in other parts of the engine as well.

Should I rename GfxPalette to GfxPalette16 anywhere (class name, property name on SciEngine)?

csnover replied:
Actually… should I also create a separate GfxPalette16 class that also includes only the palette code that is relevant to SCI16 like palVaryInstallTimer? Or maybe just replace these methods with stubs in the SCI32 subclass?…

bluegr commented:
This looks good overall, nice! :)
I've added comments in some parts that should be addressed before this gets merged in

csnover commented:
Thank you for the reviews, I will make the changes and let you know when it’s ready for another one.

csnover commented:
OK, since I did not hear anything back yet about the subclass and I am a little impatient :), I just went ahead and pulled out all the `#ifdef ENABLE_SCI32` stuff from GfxPalette into a new GfxPalette32 subclass for now. This means the old SCI palette stuff is still loaded in SCI32 engine games for the moment; over time, as more functionality gets properly RE’d and implemented from SCI32 engine games, I imagine this will end up looking more like GfxPaint.

I rebased the commits; due to the introduction of a new file GfxPalette32, the kPalCycle and kPalFade code got merged into a single commit; I don’t *think* that should be confusing in the commit history since they are explicitly written in the commit message. I did separate the commented out stub code for kSetPalStyleRange and kFrameOut into separate commits, per request.

I also discovered during rebase that an unused member property related to palette cycling was deleted from engine/state.h,cpp; I put that in a separate commit too, for clarity in history, since it was only discovered tangentially to the rest of the palette work.

This should be ready for a second review/commit now.

bluegr commented:
Looks great now, thanks!

Please, update module.mk (in the SCI engine folder) with the new file you've added before this can get merged

csnover commented:
> Please, update module.mk (in the SCI engine folder) with the new file you've added before this can get merged

Whoops. Thanks for catching that. It’s updated.

bluegr commented:
Great, thanks! Merging
